File: programmers/kakao/level4/매출_하락_최소화.py
import sys

def solution(sales, links):
    n = len(sales)
    graph = [[] for _ in range(n)]
    for a, b in links:
        graph[a-1].append(b-1)

    cost = [[0, 0] for _ in range(n)]

    def dfs(sales, node):
        cost[node][0], cost[node][1] = 0, sales[node]
        
        if not graph[node]:
            return
        
        extra = sys.maxsize
        for child in graph[node]:
            dfs(sales, child)
            if cost[child][0] < cost[child][1]:
                cost[node][0] += cost[child][0]
                cost[node][1] += cost[child][0]
                extra = min(extra, cost[child][1]-cost[child][0])
            else:
                cost[node][0] += cost[child][1]
                cost[node][1] += cost[child][1]
                extra = 0
        cost[node][0] += extra

    dfs(sales, 0)

    return min(cost[0][0], cost[0][1])


sales = [14, 17, 15, 18, 19, 14, 13, 16, 28, 17]
links = [[10, 8], [1, 9], [9, 7], [5, 4], [1, 5], [5, 10], [10, 6], [1, 3], [10, 2]]
print(solution(sales, links)) # 44

sales = [5, 6, 5, 3, 4]
links = [[2,3], [1,4], [2,5], [1,2]]
print(solution(sales, links)) # 6

sales = [5, 6, 5, 1, 4]
links = [[2,3], [1,4], [2,5], [1,2]]
print(solution(sales, links)) # 5

sales = [10, 10, 1, 1]
links = [[3,2], [4,3], [1,4]]
print(solution(sales, links)) # 2

File: Java/programmers/kakao/level4/매출하락최소화.java
package CodeTest.Java.programmers.kakao.level4;

import java.util.*;

public class 매출하락최소화 {
    private static Map<Integer, List<Integer>> graph;
    private static int[][] cost;

    public int solution(int[] sales, int[][] links) {
        int n = sales.length;

        graph = new HashMap<>();
        for (int i=0; i<n; i++) graph.put(i, new ArrayList<>());

        for (int[] link : links) {
            List<Integer> u = graph.get(link[0]-1);
            u.add(link[1]-1);
            graph.put(link[0]-1, u);
        }

        cost = new int[n][2];

        dfs(sales, 0);

        return Math.min(cost[0][0], cost[0][1]);
    }

    private static void dfs(int[] sales, int cur) {
        cost[cur][0] = 0;
        cost[cur][1] = sales[cur];

        if (graph.get(cur).isEmpty()) return;

        int extra = Integer.MAX_VALUE;

        for (Integer nxt : graph.get(cur)) {
            dfs(sales, nxt);

            if (cost[nxt][0] < cost[nxt][1]) {
                cost[cur][0] += cost[nxt][0];
                cost[cur][1] += cost[nxt][0];
                extra = Math.min(extra, cost[nxt][1]-cost[nxt][0]);
            } else {
                cost[cur][0] += cost[nxt][1];
                cost[cur][1] += cost[nxt][1];
                extra = 0;
            }
        }

        cost[cur][0] += extra;
    }
}
